src/main/java/com/microsoft/sqlserver/jdbc/SQLQueryMetadataCache.java

```diff
@@ -0,0 +1,272 @@
+            }
+        }
+        
+        session.getCryptoCache().addParamEntry(encryptionValues.getKey(), metadataMap);
+        return true;
+    }
+
+    
+    /**
+    * 
+    * Remove the cache entry.
+    * 
+    * @param stmt
+    *        SQLServer statement used to retrieve keys
+    * @param session
+    *        The enclave session where the cryptocache is stored
+    * @param connection
+    *        The SQLServerConnection, also used to retrieve keys
+    */
+    static void removeCacheEntry(SQLServerStatement stmt, EnclaveSession session,
+            SQLServerConnection connection) {
+        AbstractMap.SimpleEntry<String, String> encryptionValues = getCacheLookupKeys(stmt, connection);
+        if (encryptionValues.getKey() == null) {
+            return;
+        }
+
+        session.getCryptoCache().removeParamEntry(encryptionValues.getKey());
+    }
+
+    
+    /**
+    * 
+    * Returns the cache and enclave lookup keys for a given connection and statement
+    * 
+    * @param statement
+    *        The SQLServer statement used to construct part of the keys
+    * @param connection
+    *        The connection from which database name is retrieved
+    * @return
+    *        A key value pair containing cache lookup key and enclave lookup key
+    */
+    private static AbstractMap.SimpleEntry<String, String> getCacheLookupKeys(
+            SQLServerStatement statement, SQLServerConnection connection) {
+        final int sqlIdentifierLength = 128;
+        
+        if (connection == null) {
+            return new AbstractMap.SimpleEntry<>(null, null);
+        }
+
+        StringBuilder cacheLookupKeyBuilder = new StringBuilder();
+        cacheLookupKeyBuilder.append(":::");
+        // Pad database name to 128 characters to avoid any false cache matches because of weird DB names.
+        String databaseName = connection.activeConnectionProperties
+                .getProperty(SQLServerDriverStringProperty.DATABASE_NAME.toString());
+        cacheLookupKeyBuilder.append(databaseName);
+        for (int i = databaseName.length() - 1; i < sqlIdentifierLength; ++i) {
+            cacheLookupKeyBuilder.append(" ");
+        }
+        cacheLookupKeyBuilder.append(":::");
+        cacheLookupKeyBuilder.append(statement.toString());
+
+        String cacheLookupKey = cacheLookupKeyBuilder.toString();
+        String enclaveLookupKey = cacheLookupKeyBuilder.append(":::enclaveKeys").toString();
+
+        return new AbstractMap.SimpleEntry<>(cacheLookupKey, enclaveLookupKey);
+    }
+
+    
+    /**
+     * 
+     * Copy the enclave CEKs so they can be later used to retry secure enlave queries.
+     * 
+     * @param keysToBeSentToEnclave
+     *        The CEKs sent to the enclave cryptocache
+     * @return
+     *        A copy of the CEKs, this is what is actually added to the cryptocache
+     */
+    private static Map<Integer, CekTableEntry> copyEnclaveKeys(Map<Integer, CekTableEntry> keysToBeSentToEnclave) {

```

Comment:
remove - this is not used anywhere

Reply:
This will be used for secure enclaves in the future. Thoughts on adding an `unused` tag temporarily?